feat(article): Quacken (#386)
Nuclear-Squid/ergol: www/content/articles/la_mise_au_point_du_quacken/index.md

de ZMK.

Une autre difficulté est que le RP2040 est mal supporté par ZMK. La situation s’améliore, là encore
avec les versions 4.x de Zephyr, mais on essuie des plâtres… On remonte des bugs, l’équipe ZMK est
réactive, et le bug qu’on a relevé est corrigé dans une branche dédiée au support HWMv2 de Zephyr
(<i lang="en">hardware model v2</i>).

Tout cela fera partie de la <i lang="en">release</i> 0.4 de ZMK. Pour l’instant, on a épinglé un <i
lang="en">commit</i> bien précis pour avoir cette branche HWMv2.

### Géométrie

On a été satisfait de la géométrie dès le tout premier prototype (25.10), mais quitte à devoir faire
des nouveaux protos pour tester les corrections électroniques, on en profite pour tester des
ajustements.

Avec Ergogen, le <i lang="en">stagger</i> et le <i lang="en">splay</i> se définissent en relatif
d’un doigt à l’autre. Beaucoup de concepteurs de claviers se focalisent sur le <i
lang="en">stagger</i> de l’auriculaire, mais on s’est aperçu que c’est souvent l’*annulaire* qui
manque de <i lang="en">stagger</i> : l’écart annulaire/auriculaire est généralement suffisant.

> On a donc testé un ajustement de géométrie sur le 25.12. Un petit millimètre… mais qui fait une

Reviewer: Pédanterie : 0.1U c’est plus proche de 2 mm
```suggestion
On a donc testé un ajustement de géométrie sur le 25.12. Deux petits millimètres… mais qui fait une
```

Author: > On a donc testé un ajustement de géométrie sur le 25.12 : un dixième d’unité soit 1.7 mm… mais qui fait une différence assez nette, les annulaires et auriculaires se posent désormais *pile* au centre des touches. Adopté !

Mieux ?

Reviewer: Pan

  et le càble est trop rigide (effet de ressort entre les deux moitiés de clavier).

<i lang="en">Long story short :</i> c’est beaucoup plus simple de faire des claviers à double
contrôleur. On s’y est refusé pour des questions de coût et de complexité pour les utilisateur’ices.
Et pour une solution mono-contrôleur… le connecteur Jack 3.5 mm est le plus adapté mécaniquement,
tout en étant *acceptable* électriquement.

Un connecteur magnétique 4 points type Pogo serait idéal, mais en attendant : <i lang="en">in TRRS
ws trust</i>.

### Passer à STM32 ?

Nuke déteste ST en général et leur écosystème logiciel en particulier : basé sur Eclipse, reposant
sur de la génération de <i lang="en">boilerplate</i> plutôt que sur des abstractions logicielles,
et non libre par-dessus le marché…

Accessoirement, les puces STM32 n’ont pas de <i lang="en">bootloader</i> UF2, qui permet de flasher
par un simple glisser-déposer. Pour le Quacken, qu’on veut accessible au plus grand nombre, c’est un
problème.

Cela étant dit : si vous envisagez de faire votre propre projet embarqué (clavier ou autre), sachez
que les contrôleurs STM32 peuvent se passer d’un cristal externe (en se calant sur l’horloge du

> contrôleur USB hôte), et qu’ils intègrent leur propre mémoire Flash, anisi qu certaines résistances

Reviewer: Non, la liaison USB c’est asynchrone, d’où le fait qu’on ait besoin d’un crystal super précis (me serais beaucoup moins enmerdé si j’avais un pin d’horloge).

Les STM32F4 ont un oscillateur basé sur un montage à transistor. C’est un peu moins stable et beaucoup plus lent (à l’échelle de l’électronique, pour l’humain osef) à démarrer, mais c’est plus économique pour le fabriquant et plus simple pour intégrer la puce.

Faudrait que je rejette un œil dans les docs ST pour te faire une description plus robuste (et au passage recommander la version du STM32F4 qui a 256KB de flash au lieu de 128)

Author: Je te laisse aller voir sur Zulip mais il me semble que TeX a dit autre chose pour le cristal du STM32.

Reviewer: Autant pour moi, j’ai regardé Zulip et deux trois sources en ligne et effectivement l’oscillateur interne à l’air de se caler sur le débit des données qui arrivent du host ? Honnêtement je pige pas comment ça marche, mais ça a l’air correct du coup, je m’étais mal souvenu de ce qui a été dit, désolé pour le bruit.

Author: Merci d’avoir pris le temps de comprendre tout ça !

On a donc testé un ajustement de géométrie sur le 25.12 : un dixième d’unité soit 1.7 mm… mais qui
fait une différence assez nette, les annulaires et auriculaires se posent désormais *pile* au centre
des touches. Adopté !

Alternatives envisagées
----------------------------------------------------------------------------------------------------

### Remplacer les Jack 3.5 mm ?

En théorie, l’I²C est prévu pour être utilisé directement sur un PCB, pas sur un câble externe. En
pratique, ça se fait quand même : il y a même un standard pour ça en électronique, le [Qwiic], qui
est basé sur un connecteur JST. Peut-être que le problème, c’est juste le TRRS ?

Ayant constaté à quel point la qualité de connexion était critique pour l’I²C, se débarrasser du
TRRS paraissait logique. Les pros de l’électronique qu’on a consultés nous ont tous dit la même
chose : passer par des connecteurs Jack, c’est une erreur grossière, ils n’ont pas été faits pour ça.

On a très logiquement envisagé de remplacer les connecteurs TRRS par des Qwiic : c’est dommage de ne
plus passer par le Jack 3.5 mm qui est un peu le standard des Poticlaviers, mais si c’est meilleur…

> Problème : les connecteurs typs JST ne sont donnés que pour une trentaine de cycles de connexion /

Reviewer 2: "connecteurs typs JST" c’est une typo pour "type" ? Pas sûre de moi vu que c’est très technique ici ^^'

Author: C’est totalement une typo, je corrige, merci !

déconnexion ; et on a beau chercher, on ne trouve que deux autres connecteurs capables de supporter
les 1 000 à 10 000 cycles des connecteurs Jack.

- RJ11/RJ45 : simple, éprouvé, utilisé sur le KeyboardIO Model01, mais trop encombrant pour le
  Quacken ;

> - USB : éprouvé, mais risque de confusion avec le port USB de liaison au PC, et le càble est trop

Reviewer 2: ```suggestion
- USB : éprouvé, mais risque de confusion avec le port USB de liaison au PC, et le câble est trop
```

  rigide (effet de ressort entre les deux moitiés de clavier).

<i lang="en">Long story short :</i> ça serait beaucoup plus simple de faire des claviers à double
contrôleur, mais on s’y est refusé pour des questions de coût et de complexité pour les
utilisateur’ices. On reste donc sur une solution mono-contrôleur… et là, le connecteur Jack 3.5 mm
est le plus adapté mécaniquement, tout en étant *acceptable* électriquement.

Un connecteur magnétique 4 points type Pogo serait idéal, mais en attendant : <i lang="en">in TRRS

> ws trust</i>.

Reviewer 2: ```suggestion
we trust</i>.
```

  Quacken ;
- USB : éprouvé, mais risque de confusion avec le port USB de liaison au PC, et le câble est trop
  rigide (effet de ressort entre les deux moitiés de clavier).

<i lang="en">Long story short :</i> ça serait beaucoup plus simple de faire des claviers à double
contrôleur, mais on s’y est refusé pour des questions de coût et de complexité pour les
utilisateur’ices. On reste donc sur une solution mono-contrôleur… et là, le connecteur Jack 3.5 mm
est le plus adapté mécaniquement, tout en étant *acceptable* électriquement.

Un connecteur magnétique 4 points type Pogo serait idéal, mais en attendant : <i lang="en">in TRRS
we trust</i>.

### Passer à STM32 ?

Nuke déteste ST en général et leur écosystème logiciel en particulier : basé sur Eclipse, reposant
sur de la génération de <i lang="en">boilerplate</i> plutôt que sur des abstractions logicielles,
et non libre par-dessus le marché…

Accessoirement, les puces STM32 n’ont pas de <i lang="en">bootloader</i> [UF2], qui permet de
flasher par un simple glisser-déposer. Pour le Quacken, qu’on veut accessible au plus grand nombre,
c’est un problème.

> Cela étant dit : si vous envisagez de faire votre propre projet embarqué (clavier ou autre), sachez
> que les contrôleurs STM32 peuvent se passer d’un cristal externe (en se calant sur l’horloge du
> contrôleur USB hôte), et qu’ils intègrent leur propre mémoire Flash, anisi que certaines résistances
> et capacités de protection du circuit. C’est donc *beaucoup* plus simple à mettre en œuvre et à
> intégrer qu’un RP2040.

Author: @Nuclear-Squid cette reformulation est-elle correcte ?

```suggestion
Cela étant dit : si vous envisagez de faire votre propre projet embarqué (clavier ou autre), sachez
que les contrôleurs STM32 peuvent se passer d’un cristal externe (en se calant sur le débit de
données de l’USB), et qu’ils intègrent leur propre mémoire Flash, anisi que certaines résistances
et capacités de protection du circuit. C’est donc *beaucoup* plus simple à mettre en œuvre et à
intégrer qu’un RP2040.
```

Reviewer: Nickel !
